### PR TITLE
e2e: add case for upgrading from TiDB V4 to V5

### PR DESCRIPTION
**Pull request by handlerww**
### Code changes

- [x] Has Go code change
- [ ] Has CI related scripts change

### Tests

- [ ] Unit test 
- [x] E2E test 
- [ ] Manual test 
- [ ] No code 

### Side effects

- [ ] Breaking backward compatibility 
- [ ] Other side effects: 

### Related changes

- [ ] Need to cherry-pick to the release branch 
- [ ] Need to update the documentation 

### Release Notes

Please refer to [Release Notes Language Style Guide](https://github.com/pingcap/tidb-operator/blob/master/docs/release-note-guide.md) before writing the release note.

```release-note

```

### CODE REVIEW

**Review by DanielZhangQD — APPROVED** [review]
LGTM

**Review by apollodafoni — APPROVED** [review]

**Review by csuzhangxc — APPROVED** [review]
LGTM. BTW, should we also include other components like TiFlash and TiCDC?

**Comment by handlerww** [comment]
> LGTM. BTW, should we also include other components like TiFlash and TiCDC?

Yes, I'm going to add these components upgrading tests next week.

File: tests/e2e/tidbcluster/tidbcluster.go [hunk 1]
```diff
@@ -1929,6 +1929,45 @@ var _ = ginkgo.Describe("TiDBCluster", func() {
 			gomega.Expect(tidbCm.Data["config-file"]).To(gomega.ContainSubstring("token-limit = 10000"))
 		})
 
+		ginkgo.It("for tc and components version upgrade from TiDB V4 to TiDB V5", func() {
+			ginkgo.By("Deploy initial tc")
+			tc := fixture.GetTidbCluster(ns, "upgrade-version-v4-to-v5", utilimage.TiDBV4)
+			tc = fixture.AddTiFlashForTidbCluster(tc)
+			tc = fixture.AddTiCDCForTidbCluster(tc)
+			tc = fixture.AddPumpForTidbCluster(tc)
```
Comment (DanielZhangQD) [inline tests/e2e/tidbcluster/tidbcluster.go:1937]: Please also check the image version after upgrade.

File: tests/e2e/tidbcluster/tidbcluster.go [hunk 2]
```diff
@@ -1929,6 +1929,63 @@ var _ = ginkgo.Describe("TiDBCluster", func() {
...
+			tikvMemberName := controller.TiKVMemberName(tc.Name)
+			tikvSts, err := stsGetter.StatefulSets(ns).Get(tikvMemberName, metav1.GetOptions{})
+			framework.ExpectNoError(err, "failed to get StatefulSet %s/%s", ns, tikvMemberName)
+			tikvImage := fmt.Sprintf("pingcap/tikv:%s", componentVersion)
+			framework.ExpectEqual(tikvSts.Spec.Template.Spec.Containers[0].Image, tikvImage, "tikv sts image should be %q", tikvImage)
+
+			tiflashMemberName := controller.TiFlashMemberName(tc.Name)
+			tiflashSts, err := stsGetter.StatefulSets(ns).Get(tiflashMemberName, metav1.GetOptions{})
+			framework.ExpectNoError(err, "failed to get StatefulSet %s/%s", ns, tiflashMemberName)
+			tiflashImage := fmt.Sprintf("pingcap/tiflash:%s", componentVersion)
```
Comment (DanielZhangQD) [inline tests/e2e/tidbcluster/tidbcluster.go:1966]: `componentVersion` should be replaced with the real image, similar comments for Pump and cdc.

Comment (DanielZhangQD) [reply tests/e2e/tidbcluster/tidbcluster.go:1966]: Sorry, my mistake, no need to change this.

**Pull request merged**